components/webgpu/identity.rs

```diff
@@ -0,0 +1,85 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use crate::WebGPUMsg;
+use ipc_channel::ipc::IpcSender;
+use wgpu::{
+	hub::{IdentityHandler, IdentityHandlerFactory, GlobalIdentityHandlerFactory},
+	id::{
+        AdapterId, BindGroupId, BindGroupLayoutId, BufferId, CommandBufferId,
+        ComputePipelineId, DeviceId, PipelineLayoutId, ShaderModuleId, SwapChainId, RenderPipelineId, TextureId, TextureViewId, SamplerId, SurfaceId, TypedId
+    },
+};
+
+#[derive(Debug)]
+pub struct IdentityRecycler<I> {
+	fun: fn(I, IpcSender<WebGPUMsg>),
```

Comment:
The current approach appears to mimic Gecko code in `gfx/wgpu_bindings` very closely. I think we can make it much more elegant in Servo, since unlike Gecko we don't have to cross the FFI boundary. More concretely, I think we can get away without function pointers here altogether. Here are more details:

Instead of having  `(self.fun)(id, self.sender.clone());` for all kind of indices, we could have this implemented specifically for each kind, i.e.
```rust
impl IdentityHandler<id::DeviceId> for IdentityRecycler<id::DeviceId> {
	type Input = id::DeviceId;
	fn process(&self, id: id::DeviceId, _backend: wgpu_types::Backend) -> id::DeviceId {
		log::debug!("process {} {:?}", self.kind, id);
		//debug_assert_eq!(id.unzip().2, backend);
		id
	}
	fn free(&self, id: id::DeviceId) {
		log::debug!("free {} {:?}", self.kind, id);
                let msg = WebGPUMsg::FreeDevice(id);
                if self.sender.send(msg).is_err() {
                    log::error!("Failed to send {:?}({:?})", msg, id);
                }
	}
}
```

That would achieve the same thing without messing with any functions. We'd obviously want to generate these `impl IdentityHandler` sections by a macro, just like you are doing in this PR for `impl_factory!`.

Reply:
I'd still have to implement `IdentityHandlerFactory` for `IdentityRecyclerFactory` for each `Id` separately using `impl_factory!`, right?

Reply:
I don't thin implementing `IdentityHandlerFactory` separately for each `I` is needed. All you need is to remove the `kind` field, and it can be implemented for all `<I>` at once. The kind is only used in the logging of `IdentityHandler` implementation, where we are going to use a macro, so we can just paste the kind string directly (from a macro argument) instead of storing it as a field.

Reply:
I have been trying that, but getting stuck with specifying bounds of `I` here. Is it possible to specify `I` to be from a list of Types or are Traits the only possible way of specifying bounds?